### PR TITLE
[hivemind.Optimizer] TrainingStateAverager

### PR DESCRIPTION
**Pull request by justheuristic**
This PR implements a component of hivemind.Optimizer ( #400 ) that holds the training state and supports (delayed) optimizer steps and averaging rounds.
Unlike TrainingAverager, this class is does not need data locks as it will only update model parameters during .step.

- [x] implement TrainingStateAverager
- [x] test different initializations
- [x] test load_state_from_peers
- [x] test synchronous offloaded
- [x] test asynchronous updates (with "slow" optimizer)
- [x] include scheduler in load_state_from_peers

### CODE REVIEW

File: tests/test_optimizer.py [hunk 1]
```diff
@@ -66,3 +69,116 @@ def test_grad_averager():
...
+
+
+@pytest.mark.forked
+@pytest.mark.parametrize("offload_optimizer, reuse_tensors", [(False, False), (True, False), (False, True)])
+def test_state_averager(offload_optimizer, reuse_tensors):
+    dht1 = hivemind.DHT(start=True)
+    dht2 = hivemind.DHT(initial_peers=dht1.get_visible_maddrs(), start=True)
+
+    torch.manual_seed(1337)
+    torch.use_deterministic_algorithms(True)
```
Comment (borzunov) [inline tests/test_optimizer.py:81]: nit: This affects the execution of all further tests. Instead, we can reset this flag to the previous value in the end of this test.

File: tests/test_optimizer.py [hunk 2]
```diff
@@ -66,3 +69,116 @@ def test_grad_averager():
...
+
+    for i in range(40):
+        step, is_first_peer = i // 2, i % 2
+
+        if is_first_peer:
+            F.mse_loss(model1(x), torch.ones(3)).mul(2).backward()
+            avgr1.step(optimizer_step=True, zero_grad=True, averaging_round=(step == 10), delay_averaging=True)
+        else:
+            F.mse_loss(model2(x), -torch.ones(3)).backward()
+            avgr2.step(optimizer_step=True, zero_grad=True, averaging_round=(step == 10), delay_averaging=True)
```
Comment (borzunov) [inline tests/test_optimizer.py]:
```suggestion
    for step in range(20):
        F.mse_loss(model1(x), torch.ones(3)).mul(2).backward()
        avgr1.step(optimizer_step=True, zero_grad=True, averaging_round=(step == 10), delay_averaging=True)
        
        F.mse_loss(model2(x), -torch.ones(3)).backward()
        avgr2.step(optimizer_step=True, zero_grad=True, averaging_round=(step == 10), delay_averaging=True)
```

File: tests/test_optimizer.py [hunk 3]
```diff
@@ -66,3 +69,116 @@ def test_grad_averager():
...
+        param_groups=model2.parameters(),
+        optimizer=partial(torch.optim.Adam, lr=0.1, betas=(0.9, 0.9)),
+        scheduler=partial(torch.optim.lr_scheduler.LambdaLR, lr_lambda=lambda t: 1.0 / max(1, t)),
+        average_opt_statistics=("exp_avg_sq",),
+        offload_optimizer=offload_optimizer,
+        reuse_tensors=reuse_tensors,
+        extra_tensors=extras2,
+        target_group_size=2,
+        prefix="my_exp",
+        start=True,
```
Comment (borzunov) [inline tests/test_optimizer.py:113]: nit: Let's move `optimizer`, `scheduler`, etc. to kwargs and pass `**kwargs` to both averagers. It would be easier to see that they are the same + we'd have less code.

File: tests/test_optimizer.py [hunk 4]
```diff
@@ -66,3 +69,116 @@ def test_grad_averager():
...
+
+    avgr2 = TrainingStateAverager(
+        dht=dht2,
+        param_groups=model2.parameters(),
+        optimizer=partial(torch.optim.SGD, lr=0.1),
+        scheduler=partial(torch.optim.lr_scheduler.LambdaLR, lr_lambda=lambda t: 1.0 / max(1, t)),
+        target_group_size=2,
+        prefix="my_exp",
+        start=True,
+    )
```
Comment (borzunov) [inline tests/test_optimizer.py:175]: nit: Let's use a common kwargs dict (see above)

File: tests/test_optimizer.py (hunk 1 above)
Comment (justheuristic) [reply tests/test_optimizer.py:81]: Resolved by running test in a fork

File: tests/test_optimizer.py (hunk 2 above)
Comment (justheuristic) [reply tests/test_optimizer.py]: thanks, thats way simpler

File: tests/test_optimizer.py (hunk 4 above)
Comment (justheuristic) [reply tests/test_optimizer.py:164]: agreed

File: tests/test_optimizer.py (hunk 3 above)
Comment (justheuristic) [reply tests/test_optimizer.py]: fixd, thanks!

**Comment by justheuristic** [comment]
Note to a potential @mryab : despite running 20 optimizer steps, the test runs in ~1.3s on my laptop and ~2s in the CI.

File: hivemind/optim/experimental/state_averager.py [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+            if not delay_averaging:
+                try:
+                    output = self.pending_update.result()
+                finally:
+                    self.finished_averaging_round.clear()
+                    self.finished_optimizer_step.clear()
+        return output
+
+    def _perform_actions(self, optimizer_step: bool, zero_grad: bool, averaging_round: bool, **kwargs):
+        """Run the optimizer step, followed by a scheduler step and an averaging round, each stage is optional"""
```
Comment (borzunov) [inline hivemind/optim/experimental/state_averager.py:383]:
```suggestion
        """
        Run the optimizer step, followed by a scheduler step and an averaging round, each stage is optional.
        This method is meant to be called in the background executor.
        """
```

File: hivemind/optim/experimental/state_averager.py [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+        assert self.offload_optimizer, "Loading into offloaded optimizer requires using offloaded optimizer"
+        opt_parameters = [param for group in self.optimizer.param_groups for param in group["params"]]
+        for main_param, opt_param in zip(self._main_parameters, opt_parameters):
+            if main_param.grad is not None:
+                opt_param.grad.copy_(main_param.grad, non_blocking=True)
+
+    @torch.no_grad()
+    def _apply_optimizer_results_(self):
+        """Copy parameters from offloaded optimizer to the main model"""
+        assert self.offload_optimizer, "Applying offloaded optimizer updates requires offloaded optimizer"
```
Comment (borzunov) [inline hivemind/optim/experimental/state_averager.py:439]: nit:

```suggestion
        assert self.offload_optimizer is not None, "Applying offloaded optimizer updates requires offloaded optimizer"
```

File: hivemind/optim/experimental/state_averager.py [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+            logger.warning(f"Averaging parameters not used because averaging_round=False: {averaging_kwargs}")
+        output = None
+
+        if wait_for_delayed_update:
+            if not self.pending_update.done():
+                logger.log(self.status_loglevel, "Waiting for delayed updates to finish...")
+                output = self.pending_update.result()
+
+        if self.pending_update.done() and self.pending_update.exception():
+            logger.warning(f"Background update failed with {self.pending_update.exception()} and will be ignored.")
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/optim/experimental/state_averager.py]: resolved

File: hivemind/optim/experimental/state_averager.py [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+    for elem in flat_metadata:
+        if elem.get("type") == "tensor" and isinstance(elem.get("index"), int):
+            flat_optimizer_state.append(flat_tensors[elem["index"]])
+        elif elem.get("type") == "value" and "value" in elem:
+            flat_optimizer_state.append(elem["value"])
+    with torch.no_grad():
+        try:
+            return optimizer.load_state_dict(nested_pack(flat_optimizer_state, structure=optimizer.state_dict()))
+        except StopIteration:
+            return optimizer
```
Comment (borzunov) [inline hivemind/optim/experimental/state_averager.py:562]: Let's raise an exception instead :)

File: hivemind/optim/experimental/state_averager.py [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+
+    @staticmethod
+    def _check_params(
+        optimizer: Union[TorchOptimizer, OptimizerFactory],
+        param_groups: Optional[Union[Parameters, ParamGroups]],
+        parameter_names: Optional[Sequence[str]],
+    ) -> Tuple[ParamGroups, Sequence[torch.Tensor], Sequence[str]]:
+        """Get and verify parameters, groups and names"""
+        if (callable(optimizer) and param_groups is not None) == hasattr(optimizer, "param_groups"):
+            raise ValueError("Please provide either optimizer factory *and* param_groups or an existing optimizer")
```
Comment (borzunov) [inline hivemind/optim/experimental/state_averager.py:130]:
```suggestion
```

This seems to (1) not work with optimizer instances and (2) be already handled by the assert below.

File: hivemind/optim/experimental/state_averager.py [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+
+    def _local_tensors(self) -> Iterator[torch.Tensor]:
+        """Iterate local trainer's tensors that should be averaged with peers"""
+        for param_group in self.optimizer.param_groups:
+            yield from param_group["params"]
+        for stats in self.opt_keys_for_averaging:
+            for param_group in self.optimizer.param_groups:
+                for param in param_group["params"]:
+                    yield self.optimizer.state[param][stats]
+        yield from iter(self.extra_tensors)
```
Comment (borzunov) [inline hivemind/optim/experimental/state_averager.py:232]:
```suggestion
        yield from self.extra_tensors
```

File: hivemind/optim/experimental/state_averager.py [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+        parameters_and_extras = tuple(chain(self._main_parameters, self.extra_tensors))
+        num_parameters_and_extras = len(parameters_and_extras)
+
+        loaded_state = super().load_state_from_peers(**kwargs)
+        if loaded_state is None:
+            return
+
+        metadata, flat_tensors = loaded_state
+        if (not isinstance(metadata.get("epoch"), int)) or metadata["epoch"] < self.local_epoch:
+            logger.error("Cowardly refusing to load state from peer: peer's epoch is behind our local epoch.")
```
Comment (borzunov) [inline hivemind/optim/experimental/state_averager.py:507]:
```suggestion
            logger.warning("Cowardly refusing to load state from peer: peer's epoch is behind our local epoch")
```

nit: I prefer it to be a warning since we can continue training after that.

File: hivemind/optim/experimental/state_averager.py [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+        loaded_parameters_and_extras = flat_tensors[:num_parameters_and_extras]
+        loaded_opt_tensors = flat_tensors[num_parameters_and_extras:]
+        if num_parameters_and_extras != len(loaded_parameters_and_extras):
+            logger.error("Failed to load state from peer, received parameters, extras or metadata.")
+            return
+
+        try:
+            load_optimizer_state(self.optimizer, metadata["optimizer_metadata"], loaded_opt_tensors)
+        except StopIteration:
+            logger.error("Failed to load state from peer, received inconsistent number of optimizer statistics.")
```
Comment (borzunov) [inline hivemind/optim/experimental/state_averager.py:519]:
```suggestion
            logger.warning("Failed to load state from peer, received inconsistent number of optimizer statistics")
```

nit: I prefer it to be a warning since we can continue training after that.

**Review by borzunov — APPROVED** [review]

**Comment by justheuristic** [comment]
Note: this implements delayed optimizer step as a part of #394
